Feature couplingbeam and fluid and modified mesh morphing

cloner0110 opened the pull request:
This PR introduces two main features into the main branch:
	1.	A modified mesh morphing method
	2.	The `fvBeamPorosity` fvOption

cloner0110 committed:
initial steps to use mangrove porosity for subtracting momentum from momentum equation.

cloner0110 committed:
work in progress in coupling, draft of mangrove drag force

cloner0110 committed:
original mesh morphing replaced with modified mesh morphing

Commit:
working case with modified Mesh Morphing

cloner0110 committed:
updated fvBeamPorosity, work in progress in efficient way of defining beam properties

cloner0110 committed:
updating code, fixing some issues in fvBeamPorosity

cloner0110 committed:
updated cases based on changes on beamFoam

cloner0110 committed:
fixed some indentation within the code

cloner0110 committed:
removed unnecessary test case and minor fixes on the cases

cloner0110 committed:
Merge branch 'feature-CouplingbeamAndFluid' into feature-modifiedMeshMorphing

cloner0110 committed:
removing trailing white spaces + indentation

cloner0110 committed:
general fixes on code style

cloner0110 committed:
fixing minor coding style

cloner0110 committed:
more fixes on style

cloner0110 committed:
improving prepareBeams and cleanBeams scripts

cloner0110 committed:
some other fixes on style

cloner0110 committed:
style updates - removing modMorph comments

cloner0110 committed:
coding style fixes

cloner0110 committed:
Merge pull request #5 from solids4foam/feature-modifiedMeshMorphing

Feature modified mesh morphing

cloner0110 committed:
minor fix on equation + some fixes on style

cloner0110 committed:
immersed method is working now for low Re numbers

cloner0110 committed:
minor fix on a variable name

cloner0110 committed:
fix on the script to work on linux terminal

cloner0110 committed:
updated cases to include seabed Friction force

cloner0110 committed:
refactor fvBeamPorosity regarding coefficient calculation and include the model handling

cloner0110 committed:
fixed Smagorinsky Like formulation

cloner0110 committed:
minor fix on formulation

cloner0110 committed:
added fixed cofficient as a valid fvBeamPorosity model

cloner0110 committed:
Refactor fvBeamPorosity: Improve file handling and coefficient naming

cloner0110 committed:
Update model names

cloner0110 committed:
Test Case added - pitzDailyWithBeam

cloner0110 committed:
modified Allrun script

cloner0110 committed:
minor fix on the case

philipcardiff committed:
Update pitzDailyWithBeam to show larger beam deformation; add standard Allrun and Allclean scripts

cloner0110 committed:
fixes on Allrun and Allclean scripts

cloner0110 committed:
Merge pull request #10 from solids4foam/philipc

pitzDailyWithBeam changes

cloner0110 committed:
remove redundant cleanup commands in Allclean script

cloner0110 committed:
Updating beam properties and update restraint logic for parallel run

cloner0110 committed:
Update subproject reference in beamFoam

cloner0110 committed:
updates on fvBeamPorosity, changes on applying immersed force

cloner0110 commented:
@philipcardiff @Seevanibali @cmca01 

I’ve been working on this branch for a while, and since the modified mesh morphing could be useful in our main branch, I’m creating this PR. Previously, I had already asked for a review of the modified mesh morphing (see: https://github.com/solids4foam/moorFV/commit/0f5b5c722567e92ffbda0be35fc62d5e6117b4cc), and the code hasn’t changed since then.

The new additions are:
	•	`fvBeamPorosity`
	•	Parallel tweaks for the main library

I’ll wait for the tests to pass and look forward to your feedback.

cloner0110 committed:
Remove commented debug info and clarify description

philipcardiff commented:
FYI, @Seevanibali - please check this too

Seevanibali commented:
@philipcardiff Noted. I do not use the moorFV branch at the moment, so @cloner0110, feel free to merge it

philipcardiff committed:
Fix for Allwmakes

philipcardiff commented:
@cloner0110 : I pushed a minor edit to the Allwmake files related to finding the beamFoam directory; however, I now realise (remember) that beamFoam is added as a sub-module in moorFV, so it seems strange to expect the user to define the location of beamFoam.

Let's see... I can think of two possible "use cases":
 1. A user **does not** have a local copy of beamFoam when they download moorFV; in this case, the moorFV Allwmake script should use its local submodule version of beamFoam. This approach should be the default, as it assures that the correct branch/commit of beamFoam is used
 2. A user **does** have a local copy of beamFoam when they download moorFV; in this case, the user *may* want to use their own copy of beamFoam, or they would like moorFV to use its own version of beamFoam.

A potential solution is the following:
- If the `$BEAM_DIR` **is not** set, moorFV will download the beamFoam submodule and use it
- If the `$BEAM_DIR` **is** set, moorFV will use beamFoam located at `$BEAM_DIR` but give a warning saying that moorFV may not be compatible with that version of beamFoam. (We could actually check the git commit, but that seems like too much effort.)

I will see if I can make a quick fix for this

philipcardiff committed:
Allwmake should use beamFoam submodule by default

philipcardiff commented:
See the changes I pushed in b2b482a.

It is a bit verbose, but it seems to work.

I removed the check from `src/Allwmake` as I did not want to have to copy and paste this long piece of code and edit the local addresses.
I am sure this can be made neater and maybe a function defined (e.g. in etc/bashrc) that could be called by both scripts, but I think this is fine for now.

Once you are happy with this, I am happy for you to merge it @cloner0110

cloner0110 committed:
Immersed force calculation now works in parallel

cloner0110 committed:
merge origin/feature-CouplingbeamAndFluid

cloner0110 committed:
Now using dimensionedScalar for coefficients and simplify coefficient calculations

cloner0110 committed:
Add parallel compilation for Allwmake libraries

cloner0110 committed:
Update fvBeamPorosity to incorporate beam velocity in coefficient calculations

cloner0110 committed:
Fix coefficient calculation in fvBeamPorosity

philipcardiff commented on src/sDoFRGBFvBeam/fvOptions/fvBeamPorosity/fvBeamPorosity.C:
```diff
@@ -111,7 +112,7 @@ Foam::fv::fvBeamPorosity::coeff(const volVectorField& U, const word& modelName)
     {
         if (modelName_ == "Darcy")
         {
-            coeff = (nu_/perm_) + beta_*mag(U - beamVelocity)  * cellMarker;
```
Ah, this explains why the waves were damped when you activated this fvOption!

cloner0110 replied:
Exactly :D

cloner0110 committed:
Refactor fvBeamPorosity: Introduce calculateS method + Kernel function for ALM force

philipcardiff commented on src/sDoFRGBFvBeam/fvOptions/fvBeamPorosity/fvBeamPorosity.C:
```diff
@@ -102,11 +102,16 @@ Foam::fv::fvBeamPorosity::coeff(const volVectorField& U, const word& modelName)
         if (modelName_ == "Darcy")
         {
-            coeff = (nu_/perm_) + beta_*mag(U)  * cellMarker;
+            coeff = (nu_/perm_) + beta_*mag(U - beamVelocity)  * cellMarker;
```
Remove the space either side of `*`

:P

cloner0110 replied:
@philipcardiff nice catch :D , although I (temporarily) commented out the direct forcing approach in https://github.com/solids4foam/moorFV/pull/11/commits/e18f32a07d43ab7d91db575bbd82f678a7e65222 , and instead put the ALM methods.

philipcardiff commented on src/sDoFRGBFvBeam/fvOptions/fvBeamPorosity/fvBeamPorosity.C:
```diff
@@ -142,7 +147,7 @@ Foam::fv::fvBeamPorosity::coeff(const volVectorField& U, const word& modelName)
         if (cellMarker[c] >= 0.001 && cellMarker[c] <= 1)
         {
-            immersedForce[beamCellI] = coeff[c] * U[c] * mesh_.V()[c];
+            immersedForce[beamCellI] = coeff[c] * (U[c] - beamVelocity[c]) * mesh_.V()[c];
```
and here

cloner0110 committed:
minor style update

cloner0110 committed:
fix the force direction on the eqn.source()

cloner0110 committed:
Enhance fvBeamPorosity: Refactor calculateS method to improve beam cell coordinate handling and add force field calculations

cloner0110 committed:
fix for the compilation + commenting unused variables

cloner0110 committed:
alm force is now working in parallel

cloner0110 committed:
Implement applyBeamForce() method and refactor addSup to utilize it - Removed unused part of the code

cloner0110 committed:
sync ffvoption across all ranks

cloner0110 committed:
Enhance fvBeamPorosity: Add bounding box checks and optimize force application logic

cloner0110 committed:
Refactor fvBeamPorosity: Update beam mesh lookup and comment out bounding box checks

philipcardiff commented:
Hi @cloner0110, let me/Seevani if you want us to re-review this PR before it is merged. Or maybe you don't intend to merge it just yet.

cloner0110 commented:
@philipcardiff Sure. I'll ask for review from both of you.

cloner0110 committed:
Add README and images; enhance fvBeamPorosity with beam name and fluid density

cloner0110 commented:
## Summary

This PR adds the coupled beam-fluid interaction functionality for moorFV, extending the existing mooring-focused library toward finite-volume beam-fluid interaction applications.

Key changes include:

- Modified the mesh morphing workflow for coupled floating-body/beam simulations.
- Added an actuator-line-method style sampling and force-distribution approach through `fvBeamPorosity`.
- Added a general repository `README.md` with project overview, requirements, tutorial locations, citation, contacts, and demonstration images.

## Details

The new `fvBeamPorosity` option maps beam forces onto the fluid mesh using beam segment proximity, Gaussian kernel weighting, and normalized force distribution. The implementation now supports user-configurable beam region names, smoothing width, and fluid density through the `fvOptions` dictionary.

This PR also reorganizes the beam dependency by using `beamFoam` as a submodule rather than keeping the previous in-tree beam model sources.

cloner0110 commented:
> ## Summary
> This PR adds the coupled beam-fluid interaction functionality for moorFV, extending the existing mooring-focused library toward finite-volume beam-fluid interaction applications.
> 
> Key changes include:
> 
> * Modified the mesh morphing workflow for coupled floating-body/beam simulations.
> * Added an actuator-line-method style sampling and force-distribution approach through `fvBeamPorosity`.
> * Added a general repository `README.md` with project overview, requirements, tutorial locations, citation, contacts, and demonstration images.
> 
> ## Details
> The new `fvBeamPorosity` option maps beam forces onto the fluid mesh using beam segment proximity, Gaussian kernel weighting, and normalized force distribution. The implementation now supports user-configurable beam region names, smoothing width, and fluid density through the `fvOptions` dictionary.
> 
> This PR also reorganizes the beam dependency by using `beamFoam` as a submodule rather than keeping the previous in-tree beam model sources.

Hi @philipcardiff and @Seevanibali,

I think this PR has been open for quite some time, and my branch has started to diverge from the main branch as I continued working on it. I would really appreciate your feedback when you have a chance.

For the modified mesh morphing section, since you already reviewed it in PR #5, you can skip that part as there have been no changes since then.

FYI @cmca01.

philipcardiff commented on docs/images/p1.pdf:
Is this file used?

I think only `p1.png` is used in the `README.md`? Also, I notice that this file is much smaller than `p1.png`; I wonder if there is a way to retain the smaller file size and also the quality.

cloner0110 replied:
Fixed in https://github.com/solids4foam/moorFV/pull/11/commits/27a3cbe82efbea451d53fef8f019e1ac36eb3cea

I used codex to convert the pdf to png, It kept it alive for some reason, it is now removed.

philipcardiff commented on src/Allwmake:
```diff
@@ -1,14 +1,6 @@
-cd ${0%/*} || exit 1    # run from this directory
-
-# Check if the beamFoam installation directory has been set
-if [[ -z "${BEAMFOAM_DIR}" ]]
```
Is there a reason for removing this?

cloner0110 replied:
Thanks, basically there are two Allwmake scripts, one inside `moorFV/` and the other one inside `moorFV/src/` . Previously as you pointed out, the search for `BEAMFOAM_DIR`, was happening within the first script so I removed the checks you mentioned in the second script as it was no longer needed. In https://github.com/solids4foam/moorFV/pull/11/commits/27a3cbe82efbea451d53fef8f019e1ac36eb3cea , I fixed this issue, by explaining it clearly how to setup `BEAMFOAM_DIR`. I also have updated the `README` file for this.

cloner0110 committed:
Refactor build scripts: streamline BEAMFOAM_DIR handling and update README for clarity

Seevanibali reviewed (commented):
Maintain 80 char limit on this line - Ignore this line. I am still reviewing fvBeamPorosity.C file

Seevanibali committed:
Minor edits to fix comments of fvBeamPorosity.H file

Seevanibali commented on src/sDoFRGBFvBeam/fvOptions/fvBeamPorosity/fvBeamPorosity.H:
```diff
@@ -0,0 +1,181 @@
+{
+
+// Forward Declarations
+class MangrovesModel;
```
No object of class `MangrovesModel` is being used in the file; do we need this forward declaration?

cloner0110 replied:
Correct, MangrovesModel is not used by this class anymore. The forward declaration was leftover from the previous mangrove porosity implementation, so I removed it in https://github.com/solids4foam/moorFV/pull/11/commits/af0584b91c062cf4d39002621e1899c596f05ed2

cloner0110 committed:
Refactor fvBeamPorosity -  improve formatting

Seevanibali committed:
Comments and fixing some minor code style issues

cloner0110 committed:
update beamFoam submodule

cloner0110 committed:
Enhance Allwmake script: provide detailed instructions for beamFoam branch setup

cloner0110 committed:
Add GitHub Actions workflow for building moorFV with beamFoam submodule

cloner0110 committed:
Merge branch 'openfoam-v2306' into feature-CouplingbeamAndFluid

cloner0110 committed:
Refactor Allwmake: improve error messages for beamFoam directory checks

cloner0110 committed:
Fix beam force application in fvBeamPorosity: normalize source terms by fluid density

cloner0110 committed:
Update beamFoam submodule to latest commit

philipcardiff commented on src/sDoFRGBFvBeam/fvOptions/fvBeamPorosity/fvBeamPorosity.C:
```diff
@@ -0,0 +1,477 @@
+    \\  /    A nd           | www.openfoam.com
+     \\/     M anipulation  |
+-------------------------------------------------------------------------------
+    Copyright (C) 2017 IH-Cantabria
```
Why is this here?

cloner0110 replied:
Mangrove fvOptions are initially developed by IH-Cantabria, it is now removed in https://github.com/solids4foam/moorFV/pull/11/commits/38735fcf5970ffba203338a9fc44e5b43e314511

cloner0110 committed:
Remove outdated copyright notice in fvBeamPorosity.C

philipcardiff reviewed (approved):
This is all fine for me. Feel free to merge it
